### PR TITLE
[luci/pass] Introduce decompose_softmax

### PR DESCRIPTION
**Pull request by stamalakhov**
This draft introduces Softmax decomposing pass.

Related: #11492

ONE-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

**Comment by stamalakhov** [comment]
@jinevening 
In case `beta` for softmax needs to be supported there will be additional node for beta multiplication:
image.

I've added value tests for unit beta and non-unit beta.
Do we need to support nonunit beta?
Can i split this draft into mergeable PR's?

File: compiler/luci/pass/src/DecomposeSoftmaxPass.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+namespace luci
+{
+
+bool DecomposeSoftmaxPass::run(loco::Graph *g)
+{
+  bool changed = false;
+
+  for (auto node : loco::active_nodes(loco::output_nodes(g)))
+  {
+    if (auto hardswish = dynamic_cast<luci::CircleSoftmax *>(node))
```
Comment (jinevening) [inline compiler/luci/pass/src/DecomposeSoftmaxPass.cpp:156]: May need to rename.

File: compiler/luci/pass/src/DecomposeSoftmaxPass.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/DecomposeSoftmaxPass.h"
+
+#include "helpers/NodeFiller.h"
+#include "helpers/TypeMapper.h"
```
Comment (jinevening) [inline compiler/luci/pass/src/DecomposeSoftmaxPass.cpp:20]: Please check these helpers are in fact used.

File: compiler/luci-pass-value-test/test.lst [hunk 3]
```diff
@@ -40,6 +40,7 @@ addeval(Net_StridedSlice_StridedSlice_000 remove_unnecessary_strided_slice)
 addeval(FullyConnected_007 replace_non_const_fc_with_batch_matmul)
 addeval(Net_Transpose_Add_000 forward_transpose_op)
 addeval(Net_Transpose_Abs_000 forward_transpose_op)
+addeval(Softmax_001 decompose_softmax)
```
Comment (jinevening) [inline compiler/luci-pass-value-test/test.lst:43]: Why Softmax_002 is not included?

**Comment by jinevening** [comment]
> Do we need to support nonunit beta?

This has low priority, but it would be good to support non-unit beta now if it is not too difficult.

I think `Mul` node can be omitted if beta is 1.0.

> Can i split this draft into mergeable PR's?

Yes. The draft looks good to me. Please proceed 👍

File: compiler/luci/pass/src/DecomposeSoftmaxPass.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  beta_const->name(name + "/Softmax/beta_const");
+  luci::add_origin(beta_const, luci::get_origin(softmax));
+
+  // multiply by beta
+  auto mul = g->nodes()->create<luci::CircleMul>();
+  mul->x(sub);
+  mul->y(beta_const);
+  mul->fusedActivationFunction(luci::FusedActFunc::NONE);
+  mul->name(name + "/Softmax/beta_mul");
+  luci::add_origin(mul, luci::get_origin(softmax));
```
Comment (jinevening) [inline compiler/luci/pass/src/DecomposeSoftmaxPass.cpp:119]: If beta == 1.0, this code can be skipped.

File: compiler/luci/pass/src/DecomposeSoftmaxPass.test.cpp [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    auto graph_output = _softmax_int32_g._g.outputs()->create();
+
+    // CircleInput
+    _softmax_int32_g._input = _softmax_int32_g._g.nodes()->create<luci::CircleInput>();
+    _softmax_int32_g._input->index(graph_input->index());
+    _softmax_int32_g._input->shape({N, H, W, C});
+    _softmax_int32_g._input->dtype(loco::DataType::S32);
+    _softmax_int32_g._input->name("input");
+
+    // CircleHardSwish
```
Comment (jinevening) [inline compiler/luci/pass/src/DecomposeSoftmaxPass.test.cpp:100]:
```suggestion
    // CircleSoftmax
```

File: compiler/luci/pass/src/DecomposeSoftmaxPass.cpp (hunk 1 above)
Comment (stamalakhov) [reply compiler/luci/pass/src/DecomposeSoftmaxPass.cpp:156]: Ahhh. Sorry.

File: compiler/luci/pass/src/DecomposeSoftmaxPass.cpp (hunk 2 above)
Comment (stamalakhov) [reply compiler/luci/pass/src/DecomposeSoftmaxPass.cpp:20]: ok

File: compiler/luci-pass-value-test/test.lst (hunk 3 above)
Comment (stamalakhov) [reply compiler/luci-pass-value-test/test.lst:43]: Sorry.

**Comment by stamalakhov** [comment]
@nnfw-bot test nncc-jammy-debug

**Comment by stamalakhov** [comment]
Seems like all PR from this draft are merged, so we can close this one.

**Pull request closed without merging**